dspy/utils/cache.py

```diff
@@ -0,0 +1,123 @@
+from diskcache import FanoutCache
+from cachetools import LRUCache
+import ujson
+import pydantic
+from hashlib import sha256
+from typing import Any, Dict, List, Literal, Optional
+from functools import wraps
+from pathlib import Path
+
+class Cache:
+    """
+    dspy's caching interface. It provides 2 levels of caching (in the given order): 
+    1. An in memory cache - cachetools' lrucache 
+    2. A disk based cache - diskcache's fanoutcache
+
+    The cache is threadsafe
+
+    Args: 
+        mem_size_limit: The maximum size of the cache. If unspecified, no max size is enforced (cache is unbounded).
+        disk_size_limit: 
+    """
+    def __init__(self, directory, disk_size_limit, mem_size_limit):
+        self.memory_cache = LRUCache(maxsize=mem_size_limit)
+        self.fanout_cache = FanoutCache(shards=16, timeout=2, directory=directory, size_limit=disk_size_limit)
+        self.lock = threading.RLock()
+
+    @staticmethod
+    def cache_key(request: Dict[str, Any]) -> str:
+        """
+        Obtain a unique cache key for the given request dictionary by hashing its JSON
+        representation. For request fields having types that are known to be JSON-incompatible,
+        convert them to a JSON-serializable format before hashing.
+
+        Note: Values that cannot be converted to JSON should *not* be ignored / discarded, since
+        that would potentially lead to cache collisions. For example, consider request A
+        containing only JSON-convertible values and request B containing the same JSON-convertible
+        values in addition to one unconvertible value. Discarding the unconvertible value would
+        lead to a cache collision between requests A and B, even though they are semantically
+        different.
+        """
+
+        def transform_value(value):
+            if isinstance(value, type) and issubclass(value, pydantic.BaseModel):
+                return value.model_json_schema() # BaseModel.schema deprecated
+            elif isinstance(value, pydantic.BaseModel):
+                return value.model_dump() # BaseModel.dict deprecated
+            elif callable(value) and hasattr(value, "__code__") and hasattr(value.__code__, "co_code"):
+                return value.__code__.co_code.decode("utf-8")
+            else:
+                # Note: We don't attempt to compute a hash of the value, since the default
+                # implementation of hash() is id(), which may collide if the same memory address
+                # is reused for different objects at different times
+                return value
+
+        params = {k: transform_value(v) for k, v in request.items()}
+        return sha256(ujson.dumps(params, sort_keys=True).encode()).hexdigest()
+    
+    def get(self, request: Dict[str, Any]) -> Any:
+        try:
+            # try/except in case can't compute key
+            key = self.cache_key(request)
+        except Exception:
+            return None
+        with self.lock:  
```

Comment:
Hmm do we need to lock just to read?

Reply:
if we can remove this lock, we may wanna do something like `if key := self.memory_cache.get()` to avoid double access

Reply:
I would think reads are thread safe but the [docs](https://cachetools.readthedocs.io/en/latest/index.html#:~:text=Note%20Please%20be%20aware%20that%20all%20these%20classes%20are%20not%20thread%2Dsafe.%20Access%20to%20a%20shared%20cache%20from%20multiple%20threads%20must%20be%20properly%20synchronized%2C%20e.g.%20by%20using%20one%20of%20the%20memoizing%20decorators%20with%20a%20suitable%20lock%20object.) aren't clear.